### PR TITLE
PWN-3611 - Pin code transaction confirmation

### PR DESCRIPTION
**Pull request by kuznetsov**
### Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3611

### Description of Work

- Added `PinView` to `SendConfirmBottomSheet`
- Implemented an ability to confirm transaction by pin code
- User can make unlimited amount of pin code tries

### Screenshots

Screenshot 2022-06-03 at 17 23 49 | Screenshot 2022-06-03 at 17 23 55
:------------:|:-------------:

### CODE REVIEW

File: app/src/main/res/drawable/ic_fingerprint.xml [hunk 1]
```diff
@@ -2,10 +2,10 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     android:width="24dp"
     android:height="24dp"
-    android:tint="@color/backgroundButtonPrimary"
```
[the remark this answers is not in the document]
Comment (kuznetsov) [reply app/src/main/res/drawable/ic_fingerprint.xml:5]: Nice point, thanks

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt [hunk 2]
```diff
@@ -82,6 +93,37 @@ class SendConfirmBottomSheet(
...
+    }
+
+    private fun confirm() {
+        onConfirmed.invoke()
+        dismissAllowingStateLoss()
+    }
+
+    private fun checkPinCode(pinCode: String) {
+        CoroutineScope(dispatchers.io).launch {
+            when (authInteractor.signInByPinCode(pinCode)) {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:114]: why we use interactor inside fragment ?

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt [hunk 3]
```diff
@@ -82,6 +93,37 @@ class SendConfirmBottomSheet(
...
+        }
+    }
+
+    private fun confirm() {
+        onConfirmed.invoke()
+        dismissAllowingStateLoss()
+    }
+
+    private fun checkPinCode(pinCode: String) {
+        CoroutineScope(dispatchers.io).launch {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:113]: Redundant use of coroutines, and we should avoid launching not lifecycle aware  coroutines

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt [hunk 4]
```diff
@@ -82,6 +93,37 @@ class SendConfirmBottomSheet(
...
+            when (authInteractor.signInByPinCode(pinCode)) {
+                SignInResult.Success -> confirm()
+                SignInResult.WrongPin -> withContext(dispatchers.ui) { showWrongPinError() }
+            }
+        }
+    }
+
+    private fun showWrongPinError() {
+        val message = getString(R.string.auth_pin_code_wrong_pin)
+        context?.vibrate(VIBRATE_DURATION)
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:123]: requireContext()

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt [hunk 5]
```diff
@@ -82,6 +93,37 @@ class SendConfirmBottomSheet(
         }
     }
 
+    private fun initPinView() = with(binding) {
+        pinView.apply {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:97]: too much `this` in scopes, try rewrite using let / also

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:114]: I think we need to create presenters for BottomSheet now too

File: app/src/main/java/org/p2p/wallet/utils/BiometricPromptWrapper.kt [hunk 6]
```diff
@@ -14,18 +14,17 @@ class BiometricPromptWrapper(
...
 
         override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
-            onError?.invoke(
-                errString.takeIf {
-                    errorCode != BiometricPrompt.ERROR_USER_CANCELED &&
-                        errorCode != BiometricPrompt.ERROR_NEGATIVE_BUTTON
-                }
-            )
+            when {
+                errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON -> usePinCode?.invoke()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/utils/BiometricPromptWrapper.kt:25]: if it's a callback, maybe we should name with with prefix `on`?

**Review by gslevinkov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt (hunk 2 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:114]: `SendConfirmBottomSheet` is inherited from `NonDraggableBottomSheetDialogFragment`. There are also another bottom sheets inherited from that class and using interactors. I think we can add non draggable logic to `BaseMvpBottomSheet` and get rid of `NonDraggableBottomSheetDialogFragment` without presenters. But it looks like separate refactoring task, so let's do this outside this PR. What do you think, guys?

Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendConfirmBottomSheet.kt:114]: https://p2pvalidator.atlassian.net/browse/PWN-3831

**Review by davran1001 — APPROVED** [review]

**Review by gslevinkov — APPROVED** [review]

**Pull request closed without merging**